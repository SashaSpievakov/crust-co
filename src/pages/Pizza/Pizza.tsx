import { useEffect } from 'react';
import { useParams } from 'react-router-dom';
import ErrorRequest from 'src/components/ErrorRequest/ErrorRequest';

import FullItemBlock from '../../components/FullItemBlock/FullItemBlock';
import { Loading } from '../../components/UI';
import itemAPI from '../../services/ItemService';

export const Pizza = () => {
  const { id } = useParams();

  const { data, isLoading, isSuccess } = itemAPI.useFetchItemQuery(
    id as string,
    {
      refetchOnFocus: true,
    },
  );

  useEffect(() => {
    document.title = `Crust & Co. | Item`;
  }, []);

  return isLoading ? (
    <Loading />
  ) : isSuccess ? (
    <FullItemBlock item={data} />
  ) : (
    <ErrorRequest />
  );
};
