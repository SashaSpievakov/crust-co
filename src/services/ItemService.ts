import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

const itemAPI = createApi({
  reducerPath: 'itemAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Item'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (id) => ({
        url: `/pizzas/${id}`,
      }),
      providesTags: () => ['Item'],
    }),
  }),
});

export default itemAPI;
