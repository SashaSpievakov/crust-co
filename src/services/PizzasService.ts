import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

interface IFetchItems {
  activeCategory: number;
  sortedPropertyName: string;
}

const pizzasAPI = createApi({
  reducerPath: 'pizzasAPI',
  baseQuery: fetchBaseQuery({
    baseUrl: 'https://6344adb1dcae733e8fe3067a.mockapi.io/',
  }),
  tagTypes: ['Pizzas'],
  endpoints: (build) => ({
    fetchPizzas: build.query<IPizzaItem[], IFetchItems>({
      query: ({ activeCategory, sortedPropertyName }) => ({
        url: `pizza-items?${
          activeCategory > 0 ? `category=${activeCategory}&` : ''
        }sortBy=${sortedPropertyName}`,
      }),
      providesTags: () => ['Pizzas'],
    }),
  }),
});

export default pizzasAPI;
