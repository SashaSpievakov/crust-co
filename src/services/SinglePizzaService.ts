import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/dist/query/react';

import { IPizzaItem } from '../models/IPizzaItem';

export const pizzaAPI = createApi({
  reducerPath: 'pizzaPI',
  baseQuery: fetchBaseQuery({
    baseUrl: `${process.env.REACT_APP_API_URL}/api`,
  }),
  tagTypes: ['Pizza'],
  endpoints: (build) => ({
    fetchItem: build.query<IPizzaItem, string>({
      query: (name) => ({
        url: `/pizzas/${name}`,
      }),
      providesTags: (result) =>
        result ? [{ type: 'Pizza', id: result.name }] : ['Pizza'],
    }),
  }),
});
