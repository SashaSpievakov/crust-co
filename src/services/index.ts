export { pizzasAPI } from './PizzasService';
export { pizzaAPI } from './SinglePizzaService';
