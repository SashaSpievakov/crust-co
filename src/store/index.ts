export * from './slices';
export type { AppDispatch, AppStore, RootState } from './store';
export { setupStore } from './store';
