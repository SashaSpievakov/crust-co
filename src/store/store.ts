import {
  combineReducers,
  configureStore,
  PreloadedState,
} from '@reduxjs/toolkit';

import categorySlice from './slices/category/reducer/categoryReducer';
import sortSlice from './slices/sort/reducer/sortReducer';
import searchSlice from './slices/search/reducer/searchValueReducer';
import theme from './slices/theme/reducer/themeReducer';
import cart from './slices/cart/reducer/cartReducer';
import itemAPI from '../services/ItemService';
import pizzasAPI from '../services/PizzasService';

const rootReducer = combineReducers({
  activeCategory: categorySlice,
  activeSort: sortSlice,
  searchValue: searchSlice,
  theme,
  cart,
  [itemAPI.reducerPath]: itemAPI.reducer,
  [pizzasAPI.reducerPath]: pizzasAPI.reducer,
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware()
        .concat(itemAPI.middleware)
        .concat(pizzasAPI.middleware),
    preloadedState,
  });
};

export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];
