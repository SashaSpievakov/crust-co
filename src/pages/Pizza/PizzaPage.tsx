import { FC, useEffect } from 'react';
import { useParams } from 'react-router-dom';
import ErrorRequest from 'src/components/ErrorRequest/ErrorRequest';

import FullItemBlock from '../../components/FullItemBlock/FullItemBlock';
import { Loading } from '../../components/UI';
import { pizzaAPI } from '../../services';

export const PizzaPage: FC = () => {
  const { name } = useParams();

  const { data, isLoading, isSuccess } = pizzaAPI.useFetchItemQuery(
    name || '',
    {
      refetchOnFocus: true,
    },
  );

  useEffect(() => {
    document.title = `Crust & Co. | Item`;
  }, []);

  return isLoading ? (
    <Loading />
  ) : isSuccess ? (
    <FullItemBlock item={data} />
  ) : (
    <ErrorRequest />
  );
};
