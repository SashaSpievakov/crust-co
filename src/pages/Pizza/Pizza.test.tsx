import '@testing-library/jest-dom';

import { screen } from '@testing-library/react';
import { rest } from 'msw';

import { pizzaAPI } from '../../services';
import { setupStore } from '../../store/store';
import { renderWithProvidersAndRoutes } from '../../tests/helpers';
import { testServer } from '../../tests/mocks';
import { Pizza } from './Pizza';

describe('Pizza Page Tests', () => {
  beforeAll(() => {
    testServer.listen();
  });

  afterAll(() => {
    testServer.close();
  });

  afterEach(() => {
    testServer.resetHandlers();
    setupStore().dispatch(pizzaAPI.util.resetApiState());
  });

  test('renders the Pizza page', async () => {
    renderWithProvidersAndRoutes(null, true, '/pizzas/Chicken%20Curry');
    const itemTitle = await screen.findByText(/chicken curry/i); // eslint-disable-line @typescript-eslint/no-unused-vars

    const loadedItem = screen.getByTestId('itemPage');
    expect(loadedItem).toMatchSnapshot();
  });

  test('renders the Pizza page with a server requesr error', async () => {
    testServer.use(
      rest.get('*', (_req, res, ctx) => res.once(ctx.status(500))),
    );
    renderWithProvidersAndRoutes(<Pizza />);

    expect(
      await screen.findByText(/failed to get data from the server/i),
    ).toBeInTheDocument();
  });
});
