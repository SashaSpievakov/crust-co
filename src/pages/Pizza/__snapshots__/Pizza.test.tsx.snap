// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pizza Page Tests renders the Pizza page 1`] = `
<article
  class="sc-bcXHqe sc-fbYMXx bZXhLi hZSvIl"
  data-testid="itemPage"
>
  <picture>
    <source
      srcset="../assets/img/pizza9.webp"
      type="image/webp"
    />
    <source
      srcset="../assets/img/pizza9.png"
      type="image/png"
    />
    <img
      alt="Chicken Curry pizza"
      class="sc-GhhNo dzmXqM"
      height="450"
      src="../assets/img/pizza9.png"
      width="450"
    />
  </picture>
  <h2
    class="sc-fXqpFg bMoOTe"
  >
    Chicken Curry
  </h2>
  <p
    class="sc-dwnOUR gYzojb"
  >
    Lorem, ipsum dolor sit amet consectetur adipisicing elit. Aliquid nisi aspernatur debitis quod suscipit, vitae placeat natus ipsum inventore esse rerum animi facere numquam saepe vero mollitia quibusdam at voluptates est commodi laudantium? Iste inventore quibusdam cupiditate nemo sint iusto minus nihil culpa! Sequi architecto nesciunt explicabo mollitia. Laboriosam, odio.
  </p>
  <section
    class="sc-gGvHcT DWbJE"
  >
    <ul>
      <li
        aria-current="true"
        class="sc-ckEbSK hSXaaR"
        tabindex="0"
      >
        traditional
      </li>
      <li
        aria-current="false"
        class="sc-ckEbSK fbgxe"
        tabindex="0"
      >
        thin
      </li>
    </ul>
    <ul>
      <li
        aria-current="true"
        class="sc-ckEbSK hSXaaR"
        tabindex="0"
      >
        12
         inch
      </li>
      <li
        aria-current="false"
        class="sc-ckEbSK fbgxe"
        tabindex="0"
      >
        14
         inch
      </li>
      <li
        aria-current="false"
        class="sc-ckEbSK fbgxe"
        tabindex="0"
      >
        16
         inch
      </li>
    </ul>
  </section>
  <h3
    class="sc-UpCWa dLNuKH"
  >
    14
    $
  </h3>
  <button
    class="sc-hLBbgP sc-iBYQkv sc-pyfCe fGqcTx iDltE eflScM"
  >
    <svg
      class="sc-laZRCg ekmfhU"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      stroke-width="0"
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M8 0a1 1 0 0 1 1 1v6h6a1 1 0 1 1 0 2H9v6a1 1 0 1 1-2 0V9H1a1 1 0 0 1 0-2h6V1a1 1 0 0 1 1-1z"
      />
    </svg>
    <span>
      Add
    </span>
  </button>
  <a
    class="sc-hLBbgP sc-eDvSVe fGqcTx lhIsum"
    href="/"
  >
    Go Back
  </a>
</article>
`;
